### PR TITLE
Change libsvm classifier to use scikit-learn svm Implementation

### PR DESCRIPTION
**Pull request by joshanderson-kw**

### CODE REVIEW

**Comment by Purg** [comment]
Before I get into the diff, I should point out that we should add the scikit-learn SVM plugin as a *separate*, *new* plugin, not a replacement over the existing one. This way we can retain the "old" implementation if a use-case or environment finds it specifically applicable, as well as some of the changes here are specifically to match the current implementation's behavior and side-effects even though they don't really apply to scikit-learn.

This pivot might also simplify your other task of an "equivalency check" notebook.

File: smqtk_classifier/impls/classify_descriptor_supervised/sklearn_svm.py [hunk 1]
```diff
@@ -0,0 +1,449 @@
...
+        svm_label_map_uri: Optional[str] = None,
+        train_params: SVM_PARAM_MAPPING_T = {
+            '-s': 0,  # C-SVC, assumed default if not provided
+            '-t': 0,  # linear kernel
+            '-b': 1,  # enable probability estimates
+            '-c': 2,  # SVM parameter C
+        },
+        normalize: Optional[Union[int, float, str]] = None,
+        n_jobs: Optional[int] = 4,
+    ):
```
Comment (Purg) [inline smqtk_classifier/impls/classify_descriptor_supervised/sklearn_svm.py:91]: We can treat this as a "new" implementation: a.k.a. we don't need to mirror the libSVM's constructor etc. so exactly here. Let's keep it relevant to what *this* implementation wants, e.g. `train_params` was for libSVM, we can be more direct here for the scikit-learn requirements. More e.g. we could stick to this impl wrapping the `svm.SVC` algorithm for now; I don't think the other SVM variants were ever used.

File: smqtk_classifier/impls/classify_descriptor_supervised/sklearn_svm.py [hunk 2]
```diff
@@ -0,0 +1,449 @@
...
+            "Should have an SVM model at this point."
+        )
+
+        # Dump descriptors into a matrix for normalization and use in
+        # prediction.
+        vec_mat = numpy.array(list(array_iter))
+        vec_mat = self._norm_vector(vec_mat)
+        n_jobs = self.n_jobs
+        if n_jobs is not None:
+            n_jobs = min(len(vec_mat), n_jobs)
```
Comment (Purg) [inline smqtk_classifier/impls/classify_descriptor_supervised/sklearn_svm.py:390]: I don't think the previous parallelism model holds for the now-use of sklearn's SVC, whose predict_proba function I believe is vectorized which makes parallelism make little to no sense here.

File: smqtk_classifier/impls/classify_descriptor_supervised/sklearn_svm.py [hunk 3]
```diff
@@ -0,0 +1,449 @@
...
+    def get_config(self) -> Dict[str, Any]:
+        return {
+            "svm_model_uri": self.svm_model_uri,
+            "svm_label_map_uri": self.svm_label_map_uri,
+            "train_params": self.train_params,
+            "normalize": self.normalize,
+            "n_jobs": self.n_jobs,
+        }
+
+    def __getstate__(self) -> Any:
```
Comment (Purg) [inline smqtk_classifier/impls/classify_descriptor_supervised/sklearn_svm.py:133]: The use of this an `__setstate__` are likely not needed in this case due not wrapping a C-library's handle pointer any more. I assume the wrapped scikit-learn model instance knows how to un/pickle itself.

File: smqtk_classifier/impls/classify_descriptor_supervised/sklearn_svm.py [hunk 4]
```diff
@@ -0,0 +1,449 @@
...
+        if not self.has_model():
+            raise RuntimeError("No model loaded")
+        return list(self.svm_label_map.values())
+
+    def _classify_arrays(self, array_iter: Union[np.ndarray, Iterable[np.ndarray]]) -> Iterator[Dict[Hashable, float]]:
+        if not self.has_model():
+            raise RuntimeError("No SVM model present for classification")
+        assert self.svm_model is not None, (
+            "Should have an SVM model at this point."
+        )
```
Comment (Purg) [inline smqtk_classifier/impls/classify_descriptor_supervised/sklearn_svm.py:382]: Redundant with the above check?

File: smqtk_classifier/impls/classify_descriptor_supervised/sklearn_svm.py [hunk 5]
```diff
@@ -0,0 +1,449 @@
...
+        param_debug = {'-q': ''}
+        if LOG.getEffectiveLevel() <= logging.DEBUG:
+            param_debug = {}
+
+        # Form libSVM problem input values
+        LOG.debug("Formatting problem input")
+        train_labels = []
+        train_vectors = []
+        train_group_sizes = []  # number of examples per class
+        self.svm_label_map = {}
```
Comment (Purg) [inline smqtk_classifier/impls/classify_descriptor_supervised/sklearn_svm.py:271]: If I understand correctly, we should be able to pass labels, even as strings, directly to the scikit-learn Estimator `fit` function, and it will also record those labels in the `classes_` instance attribute, so storing those labels here is redundant.

**Review by Purg — COMMENTED** [review]
TLDR: This new impl. could likely be massively simplified by not following so exactly the implementation of the libSVM as a number of aspects do not make sense to continue (see more specific comments).

**Comment by joshanderson-kw** [comment]
> TLDR: This new impl. could likely be massively simplified by not following so exactly the implementation of the libSVM as a number of aspects do not make sense to continue (see more specific comments).

I removed quite a bit of the libSVM stuff. Hopefully not too much. It still works with the IQR demo at least.

**Review by Purg — APPROVED** [review]
